### PR TITLE
refactor: remove the phase concept and lean in to labels

### PR DESCRIPTION
**Pull request by GeorgeMac**
This replaces: https://github.com/get-glu/glu/pull/6

This is a large refactor to all the things. Primarilly, it removes the concept of phases.
The experiment binary example is a good place to start to see how the world now looks:
```go
func run(ctx context.Context) error {
	return glu.NewSystem(ctx).AddPipeline(func(ctx context.Context, config *glu.Config) (glu.Pipeline, error) {
		// fetch the configured OCI repositority source named "checkout"
		ociRepo, err := config.OCIRepository("checkout")
		if err != nil {
			return nil, err
		}

		ociSource := oci.New[*CheckoutResource](ociRepo)

		// fetch the configured Git repository source named "checkout"
		gitRepo, gitProposer, err := config.GitRepository(ctx, "checkout")
		if err != nil {
			return nil, err
		}

		gitSource := git.NewSource[*CheckoutResource](gitRepo, gitProposer, git.ProposeChanges, git.AutoMerge)

		// create initial (empty) pipeline
		pipeline := glu.NewPipeline(glu.Name("checkout"), NewCheckoutResource)

		// build a controller which sources from the OCI repository
		ociController := controllers.New(glu.Name("oci"), pipeline, ociSource)

		// build a controller for the staging environment which source from the git repository
		// configure it to promote from the OCI controller
		gitStaging := controllers.New(glu.Name("git-staging", glu.Label("env", "staging")),
			pipeline, gitSource, core.PromotesFrom(ociController))

		// build a controller for the production environment which source from the git repository
		// configure it to promote from the staging git controller
		_ = controllers.New(glu.Name("git-production", glu.Label("env", "production")),
			pipeline, gitSource, core.PromotesFrom(gitStaging))

		// return configured pipeline to the system
		return pipeline, nil
	}).ScheduleReconcile(
		glu.ScheduleInterval(10*time.Second),
		glu.ScheduleMatchesLabel("env", "staging"),
		// alternatively, the controller instance can be target directly with:
		// glu.ScheduleMatchesController(gitStaging),
	).Run()
}
```

Now that phases are gone, the system leans into `labels` a lot more.
Instead of registries, pipelines, phases, reconcilers and resources, we now have system, pipelines, controllers and resources.

- `System` replaces `Registry` and now manages schedules itself. Schedules can be applied to match specific controllers or label sets.
- `Pipeline` is now type to the `Resource` that flows through it. You must define N pipelines if you want to flow different resources types through the entire system.
- `DependsOn` has been renamed `PromotesFrom` to make the intention clearer.
- `Controllers` replaces the previously `Reconciler` concept. Reconciler was an interface and Controller is a now struct. It can take source and encapsulates the common logic for decided whether or not to promote from a dependent controller.
- `Resource` types no longer have their own metadata. However, they're presented with the acting controllers metadata on reconcile. This allows the controller to pass down details such as its own labels (for e.g. scoping operations to a particular env or phase).

### CODE REVIEW

File: pipeline.go [hunk 1]
```diff
@@ -0,0 +1 @@
+package glu
```
Comment (markphelps) [inline pipeline.go:1]: do we need this file still?

File: server.go [hunk 2]
```diff
@@ -39,8 +39,7 @@ func (s *Server) setupRoutes() {
 	s.router.Route("/api/v1", func(r chi.Router) {
 		r.Get("/pipelines", s.listPipelines)
 		r.Get("/pipelines/{pipeline}", s.getPipeline)
-		r.Get("/pipelines/{pipeline}/phases/{phase}", s.getPhase)
-		r.Get("/pipelines/{pipeline}/phases/{phase}/resources/{resource}", s.getResource)
+		r.Get("/pipelines/{pipeline}/controller/{phase}", s.getController)
```
Comment (markphelps) [inline server.go:42]:
```suggestion
		r.Get("/pipelines/{pipeline}/controller/{controller}", s.getController)
```

**Review by markphelps — CHANGES_REQUESTED** [review]
one change, one q. otherwise lgtm!

File: pipeline.go (hunk 1 above)
Comment (GeorgeMac) [reply pipeline.go:1]: Indeed!

**Review by markphelps — APPROVED** [review]

**Pull request merged**